pkg/migration-tool/reader/path_reader.go

```diff
@@ -0,0 +1,169 @@
+func NewPathRead(c context.Context, path string, p *plan.Plan, sigRead chan *plan.Block, needsStopChan bool) (Reader, error) {
+	reader := &PathRead{
+		c:            c,
+		path:         path,
+		plan:         p,
+		sigBlockRead: sigRead,
+	}
+	if needsStopChan {
+		reader.sigForceStop = make(chan struct{})
+	}
+	return reader, nil
+}
+
+func (rr *PathRead) SigStop() {
+	rr.sigForceStop <- struct{}{}
+}
+
+// Run runs the remote read and starts fetching the samples from the read storage.
+func (rr *PathRead) Run(errChan chan<- error) {
+	var err error
+	db, err := tsdb.OpenDBReadOnly(rr.path, nil)
+	if err != nil {
+		errChan <- fmt.Errorf("path-read run: opening prom tsdb: %w", err)
+		return
+	}
+	querier, err := db.Querier(rr.c, rr.plan.Mint(), rr.plan.Maxt())
+	if err != nil {
+		errChan <- fmt.Errorf("path-read run: prom tsdb querier: %w", err)
+		return
+	}
+	ss := querier.Select(false, nil, labels.MustNewMatcher(labels.MatchRegexp, "", ".*"))
+	type itrMap struct {
+		labels labels.Labels
+		itr    chunkenc.Iterator
+	}
+	var iteratorStore = make(map[string]itrMap)
+	// Fill the iteratorStore with the iterators across all the series-sets for the given time-range.
+	for ss.Next() {
+		series := ss.At()
+		iteratorStore[series.Labels().String()] = itrMap{labels: series.Labels(), itr: series.Iterator()}
+	}
+	go func() {
+		defer func() {
+			if err = tsdb_errors.NewMulti(err, db.Close()).Err(); err != nil {
+				errChan <- fmt.Errorf("path-read run: %w", err)
+			}
+			if err = querier.Close(); err != nil {
+				errChan <- fmt.Errorf("path-read run: %w", err)
+			}
+			close(rr.sigBlockRead)
+			log.Info("msg", "reader is down")
+			close(errChan)
+		}()
+		log.Info("msg", "reader is up")
+		// Start creating blocks and filling them with samples.
+		for rr.plan.ShouldProceed() {
+			select {
+			case <-rr.c.Done():
+				return
+			case <-rr.sigForceStop:
+				return
+			default:
+			}
+			blockRef, err := rr.plan.NextBlock()
+			if err != nil {
+				errChan <- fmt.Errorf("path-run run: %w", err)
+				return
+			}
+			ts := make(tsMap)
+			mint, maxt := blockRef.Mint(), blockRef.Maxt()
+			populateTimeseries := func(m itrMap) bool {
+				// Populate the timeseries map with the iterators for the current time-range delta.
+				// We maintain a local store of series iterators in iteratorStore. This keeps the last progress in the memory
+				// (at a negligible cost since iterator), leading to faster access of samples for subsequent time-ranges.
+				itr := m.itr
+				var samples []prompb.Sample
+				appender := func() bool {
+					t, v := itr.At()
+					if t >= mint && t < maxt {
+						samples = append(samples, prompb.Sample{
+							Timestamp: t,
+							Value:     v,
+						})
+						return true
+					}
+					return false
+				}
+				if itr.Seek(mint) {
+					appender()
+					for itr.Next() {
+						if !appender() {
+							break
+						}
+					}
+				}
+				if len(samples) > 0 {
+					ts[m.labels.String()] = struct {
```

Comment:
This can be a slice of []*prompb.Timeseries or [][]byte(see below). That way we avoid a tsMapToTs and and intermediate representation.